File: firestarter/src/main/java/com/googlecode/blaisemath/firestarter/editor/ChooserComboPopup.java
package com.googlecode.blaisemath.firestarter.editor;

/*
 * #%L
 * Firestarter
 * --
 * Copyright (C) 2009 - 2025 Elisha Peterson
 * --
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import java.awt.BorderLayout;
import java.awt.Color;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import javax.swing.JButton;
import javax.swing.JColorChooser;
import javax.swing.JPanel;
import javax.swing.JPopupMenu;

public class ChooserComboPopup extends JPopupMenu {

    private final ColorEditor ce;

    public ChooserComboPopup(ColorEditor c) {
        super();
        this.ce = c;
        JPanel p = new JPanel();
        p.setLayout(new BorderLayout());
        final SwatchChooserPanel s = new SwatchChooserPanel(c, this);
        s.buildChooser();
        p.add(s, BorderLayout.NORTH);
        JButton b = new JButton("Other ...");
        b.addActionListener(e -> {
            Color color = JColorChooser.showDialog(getParent(), "Color Chooser", null);
            ce.setNewValue(color);
            ce.initEditorValue();
            setVisible(false);
        });
        p.add(b, BorderLayout.SOUTH);
        add(p);
        s.addMouseListener(new MouseAdapter() {
            @Override
            public void mouseReleased(MouseEvent e) {
                setVisible(false);
            }
        });
    }

    public ColorEditor getCE() {
        return this.ce;
    }
}

File: firestarter/src/main/java/com/googlecode/blaisemath/firestarter/editor/ChooserComboPopup.kt
package com.googlecode.blaisemath.firestarter.editor

import com.googlecode.blaisemath.app.ApplicationMenuConfig
import com.googlecode.blaisemath.firestarter.editor.EditorRegistrationTest
import com.googlecode.blaisemath.firestarter.property.BeanTreeNode
import com.googlecode.blaisemath.firestarter.property.CustomBean
import com.googlecode.blaisemath.firestarter.property.HelloWorldTestFrame
import com.googlecode.blaisemath.firestarter.property.IndexedBean
import com.googlecode.blaisemath.firestarter.property.IndexedBean.Indexed2
import com.googlecode.blaisemath.firestarter.property.IndexedBean.TestEnum
import com.googlecode.blaisemath.firestarter.property.NumberBean
import com.googlecode.blaisemath.firestarter.property.PointBean
import com.googlecode.blaisemath.firestarter.property.PropertySheetDialogTestFrame
import com.googlecode.blaisemath.firestarter.property.ShapeBean
import com.googlecode.blaisemath.firestarter.property.TestPropertyModel
import com.googlecode.blaisemath.firestarter.property.Testing
import com.googlecode.blaisemath.ui.PropertyActionPanel
import junit.framework.TestCase
import org.apache.commons.math.stat.descriptive.SummaryStatistics
import org.junit.Before
import org.junit.BeforeClass
import java.awt.BorderLayout
import java.awt.event.ActionEvent
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import javax.swing.JButton
import javax.swing.JColorChooser
import javax.swing.JPanel
import javax.swing.JPopupMenu

/*
* #%L
* Firestarter
* --
* Copyright (C) 2009 - 2019 Elisha Peterson
* --
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*      http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
* #L%
*/   class ChooserComboPopup(private val ce: ColorEditor?) : JPopupMenu() {
    fun getCE(): ColorEditor? {
        return ce
    }

    init {
        val p = JPanel()
        p.layout = BorderLayout()
        val s = SwatchChooserPanel(ce, this)
        s.buildChooser()
        p.add(s, BorderLayout.NORTH)
        val b = JButton("Other ...")
        b.addActionListener { e: ActionEvent? ->
            val color = JColorChooser.showDialog(parent, "Color Chooser", null)
            ce.setNewValue(color)
            ce.initEditorValue()
            isVisible = false
        }
        p.add(b, BorderLayout.SOUTH)
        add(p)
        s.addMouseListener(object : MouseAdapter() {
            override fun mouseReleased(e: MouseEvent?) {
                isVisible = false
            }
        })
    }
}